src/Castle.Core/DynamicProxy/IInvocationProceedInfo.cs

```diff
@@ -0,0 +1,31 @@
+// Copyright 2004-2019 Castle Project - http://www.castleproject.org/
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.DynamicProxy
+{
+	using System;
+
+	/// <summary>
+	///   Describes the <see cref="IInvocation.Proceed"/> operation for an <see cref="IInvocation"/>
+	///   at a specific point during interception.
+	/// </summary>
+	public interface IInvocationProceedInfo
+	{
+		/// <summary>
+		///   Executes the <see cref="IInvocation.Proceed"/> operation described by this instance.
+		/// </summary>
+		/// <exception cref="NotImplementedException">There is no interceptor, nor a proxy target object, to proceed to.</exception>
+		void Invoke();
```

Comment:
Is there a reason you chose to call this `Invoke` rather than `Proceed`? Is it just because you had `GetProceedInfo` returning a delegate?

Reply:
There's two aspects to this:

* Linguistics. Given that `IInvocationProceedInfo` represents a specific `Proceed` call, `Invoke` seemed like a more appropriate name: You can "invoke" the `Proceed` operation; you don't "proceed to" the `Proceed` operation.

* Extensibility. We could just as well have a method `Action IInvocation.GetProceeder()` that returns a simple delegate. I opted for something more extensible. Originally I had additional methods on `IInvocationProceedInfo` that allowed inspection of what exactly you'd be proceeding to: `bool IsInterceptor(out IInterceptor)` and `bool IsTarget(out object)`. It turned out that those aren't needed to implement an async interceptor so I removed them, but I liked the idea of being able to "reflect over" the interception pipeline, so I kept `IInvocationProceedInfo` instead of a simple delegate.

That being said, if you prefer the method to be named `Proceed`, I don't really mind renaming it.

Reply:
Makes sense. This object obviously describes where you are proceeding to. I was thinking more like this object represented the current context so you'd "proceed" from the current position/index in the chain since that is what it stores. Semantics I guess, you've been consistent with naming so `Invoke` fits.